iceoryx_hoofs/cmake/IceoryxPlatform.cmake

```diff
@@ -44,6 +44,20 @@ elseif(CMAKE_CXX_COMPILER_ID MATCHES "GNU" OR CMAKE_CXX_COMPILER_ID MATCHES "Cla
     set(ICEORYX_WARNINGS PRIVATE ${ICEORYX_WARNINGS} -W -Wall -Wextra -Wuninitialized -Wpedantic -Wstrict-aliasing -Wcast-align -Wno-noexcept-type -Wconversion)
 endif()
 
+## ignore -Wmaybe-uninitialized for gcc
+## From the gcc documentation:
+##    compiler emits a warning if it cannot prove the uninitialized paths are 
+##    not executed at run time.
+## Here are very hard to debug false positives possible when the compiler is unable
+## to prove that a certain variable is initialized. This warning does not mean
+## that the variable is used uninitialized. It states that the variable is `maybe`
+## used uninitialized.
+## In combination with -Werror this leads to compile error and the only fix is
+## to assign already initialized variables again a value which can be very costly.
+if(CMAKE_CXX_COMPILER_ID MATCHES "GNU")
+    set(ICEORYX_WARNINGS PRIVATE ${ICEORYX_WARNINGS} -Wno-maybe-uninitialized)
+endif()
```

Comment:
I downloaded the gcc 11.2 sources (latest release) and built them on ubuntu 20.04 but I don't get any warning when I remove these lines

Reply:
... and after disabling the sanitizer I also get the error but only for the introspection client ... something is fishy with the client. I'll investigate that

Reply:
Okay, I have a minimal example
https://github.com/ApexAI/iceoryx/blob/weird-maybe-uninitialized-warning/tools/introspection/source/introspection_app.cpp#L44:L68
While disabling the warning for iceoryx, this doesn't fix the problem since someone else could build with this warning enabled and encounter the same issue.

We either have to
1. understand why this happens
2. suppress the warning where it occurs
3. disable the warning in the introspection client itself and let all our user do the same if they encounter the problem

I'm for 1 and/or 2.

Reply:
This would be a "fix" ... at least there is no warning anymore, even though it was most probably a false positive
https://github.com/ApexAI/iceoryx/commit/174f9bee476ba53e3a5241bfc05090d7e0c81acc

Reply:
@elBoberido I am against 1. The reason is that we actually do not suppress the cause of that error but symptoms.

My investigation showed that the source of the error is an uninitialized `m_data` member in the `optional` but the error occurred in some lines of the `function_ref` which was contained in the optional.

So we would suppress the error in the `function_ref` which is not the cause of the error and when you think longer about this it is not unlikely that we have other constructs which are contained in some kind of `optional` which also encounter `maybe-uninitialized` warnings.
So we ignore the warning then all over the place but not in this one single origin which causes it.

We also have a third option:
3. Initialize `m_data` in the `optional` with zero.

But I am against this one. Just think about the overhead when the object inside the optional is huge then we have a useless initialization of memory just to fix a broken compiler warning.

I would vote for 2.

Reply:
@elBoberido 
> This would be a "fix" ... at least there is no warning anymore, even though it was most probably a false positive
ApexAI@174f9be

I tried it out and also adapted in for the `or_else` path since we potentially have the same issue there. It fixed the issue so really thank you for your effort! This helped a lot and is definitely the cleanest and best solution.